Cache name resolve results in astropy.coordinates

adrn opened the pull request:
This PR adds support for caching results from the name resolve (Sesame) functionality in astropy.coordinates (primarily used via `SkyCoord.from_name`).

The motivator here was that, in astropy-tutorials (and in my own scripts), we use `.from_name()` to retrieve coordinates of some objects every time the build is run, but this unnecessarily hammers the Sesame servers and occasionally fails because of connection issues. 

Before I add a changelog entry, I figured I would open the PR to see if anyone has strong opinions about this.

eteq commented:
I wanted to say this seems dangerous... But honestly when I think on it I realize this would indeed be super useful for me in my own science some of the time - often I grumble that I can't keep going with something using `from_name` when the internet goes out. So generally I like it (haven't reviewed in detail, but that's my first reaction).

I do think it needs a way to bust the cache easily though. I.e., "I know sesame was updated last week and I want the newest coordinates". Right now that requires knowing how to clear the cache via a url the user never sees... Maybe just have an option for `cache` to be `'overwrite'` or similar)

adrn commented:
Good point! I added an `overwrite` argument - I personally don't like having boolean arguments that also accept magic string values, so I opted for another argument instead.

aarchiba commented:
Why is it necessary to manually handle locking and caching? Would astropy.utils.data.download_file do what you want if it had an `overwrite` argument?

aarchiba commented on astropy/coordinates/name_resolve.py:
```diff
@@ -9,17 +9,24 @@
 from astropy import units as u
 from .sky_coordinate import SkyCoord
 from astropy.utils import data
+from astropy.utils.data import (_get_download_cache_locs,
```
It's really better not to use internal functions like this if there's an atlernative. At a minimum you're moving implementation details outside astropy.utils.data.

pllim replied:
Yes, this will be impacted by #9182

aarchiba commented on astropy/coordinates/name_resolve.py:
```diff
@@ -152,12 +167,58 @@ def get_icrs_coordinates(name, parse=False):
     for url in urls:
+
+        if cache and not overwrite:
+            with shelve.open(urlmapfn) as url2hash:
```
It is potentially bug-prone not to lock on read. Maybe it's fine. But better to leave that kind of worry inside astropy.utils.data.

aarchiba commented on astropy/coordinates/name_resolve.py:
```diff
@@ -152,12 +167,58 @@ def get_icrs_coordinates(name, parse=False):
+
+                        local_path = os.path.join(dldir, hash.hexdigest())
+                        with open(local_path, 'wb') as f:
+                            f.write(resp_data)
```
This entire loop body could be replaced by `open(astropy.utils.download_file(url, cache=cache, overwrite=overwrite), "rb").read()` if download_file supported `overwrite`. That would be a fairly easy addition, particularly once #9182 is merged.

mhvk commented:
Like the idea and wonder if we can at least make a session-local cache the default...

But agree with @aarchiba that the actual cache handling should be in `utils.data`, not here.

bsipocz commented:
Needs a rebase and review being addressed. Most likely also a milestone change.

adrn commented:
I rebased and switched to using the new `download_file` functionality. However, this currently doesn't work because of the way CDS delivers data, and they way we generate hashes for the cache. Each response from CDS may not be identical because it includes (amongst other things) the amount of time the query took and the date/time, even if repeatedly querying with the same search string. For example, here are two responses with the same query "Castor":

```
# castor	#Q62208733
#!Vl=VizieR (local): No table found for: castor

#(Q62208730)

#=Sa=Simbad@CfA:    1     0ms (from cache)
%@ 983633
%I.0 * alf Gem
%C.0 **
%C.N0 12.13.00.00
%J 113.649471640 +31.888282216 = 07:34:35.87 +31:53:17.8
%J.E [34.72 25.95 90] A 2007A&A...474..653V
%P -191.45 -145.19 [3.95 2.95 0] A 2007A&A...474..653V
%X 64.12 [3.75] A 2007A&A...474..653V
%V v 5.40 ~ [0.5] A 2006AstL...32..759G
%S A1V+A2Vm =0.0000D200.0030.0110000000100000 E ~
%#B 235


#!Vl=VizieR (local): No table found for: castor

#(Q62208730)

#=Sa=Simbad@CfA:    1     0ms (from cache)
%@ 983633
%I.0 * alf Gem
%C.0 **
%C.N0 12.13.00.00
%J 113.649471640 +31.888282216 = 07:34:35.87 +31:53:17.8
%J.E [34.72 25.95 90] A 2007A&A...474..653V
%P -191.45 -145.19 [3.95 2.95 0] A 2007A&A...474..653V
%X 64.12 [3.75] A 2007A&A...474..653V
%V v 5.40 ~ [0.5] A 2006AstL...32..759G
%S A1V+A2Vm =0.0000D200.0030.0110000000100000 E ~
%#B 235



#====Done (2019-Dec-11,18:16:23z)====
```

and

```
# castor	#Q62208730
#!Vl=VizieR (local): No table found for: castor
#=Sa=Simbad@CfA:    1    52ms
%@ 983633
%I.0 * alf Gem
%C.0 **
%C.N0 12.13.00.00
%J 113.649471640 +31.888282216 = 07:34:35.87 +31:53:17.8
%J.E [34.72 25.95 90] A 2007A&A...474..653V
%P -191.45 -145.19 [3.95 2.95 0] A 2007A&A...474..653V
%X 64.12 [3.75] A 2007A&A...474..653V
%V v 5.40 ~ [0.5] A 2006AstL...32..759G
%S A1V+A2Vm =0.0000D200.0030.0110000000100000 E ~
%#B 235
#!Vl=VizieR (local): No table found for: castor

#(Q62208730)

#=Sa=Simbad@CfA:    1    14ms
%@ 983633
%I.0 * alf Gem
%C.0 **
%C.N0 12.13.00.00
%J 113.649471640 +31.888282216 = 07:34:35.87 +31:53:17.8
%J.E [34.72 25.95 90] A 2007A&A...474..653V
%P -191.45 -145.19 [3.95 2.95 0] A 2007A&A...474..653V
%X 64.12 [3.75] A 2007A&A...474..653V
%V v 5.40 ~ [0.5] A 2006AstL...32..759G
%S A1V+A2Vm =0.0000D200.0030.0110000000100000 E ~
%#B 235

#====Done (2019-Dec-11,18:16:23z)====
```

In the previous set of commits (overwritten with the rebase), this is why I was manually interfacing with the cache: I needed to set the hash to the URL instead of the file content. 

So, in order to get this PR to work, we would need to add functionality to `download_file` and related machinery to specify or overwrite the hash...

pllim commented:
I think you already can. See "If they are concerned that the primary source of the data..." in https://astropy.readthedocs.io/en/latest/utils/data.html

aarchiba commented:
> I rebased and switched to using the new `download_file` functionality. However, this currently doesn't work because of the way CDS delivers data, and they way we generate hashes for the cache. Each response from CDS may not be identical because it includes (amongst other things) the amount of time the query took and the date/time, even if repeatedly querying with the same search string. For example, here are two responses with the same query "Castor":

Indeed, the result is going to be different every time, and if the cache works, you will get the same result as the first successful try (unless you choose to use cache="update" and re-download). Why is this a problem?

> In the previous set of commits (overwritten with the rebase), this is why I was manually interfacing with the cache: I needed to set the hash to the URL instead of the file content.

Why? The hash need not be user-visible.

> So, in order to get this PR to work, we would need to add functionality to `download_file` and related machinery to specify or overwrite the hash...

Why? I don't understand.

If I understand what you want (maybe I don't?) you should be able to write:

```python
response = get_file_contents(download_file("http://cds.query/goes?here", cache=True))
```

This will get you a fresh request the first time and a copy of that result every later time. Isn't that what you want? The hash value appears only in the filename returned from `download_file` and the user never needs to care about it. (I considered just replacing the hash with a UUID in this context).

astrofrog commented:
This seems worthwhile, so just wanted to give a :+1: on the idea!

adrn commented:
Oh right, this PR! Updated and rebased.

adrn commented:
RE: @aarchiba's comments - I don't remember why I was confused before. On revisiting this, it seemed pretty straightforward to get this working, so I think the new commits address the remaining issues. The test right now is pretty "weak," so open to suggestions there, but don't think it's critical.

adrn commented:
OK, sure!

pllim commented on astropy/coordinates/name_resolve.py:
```diff
@@ -155,9 +160,9 @@ def get_icrs_coordinates(name, parse=False):
-            resp = urllib.request.urlopen(url, timeout=data.conf.remote_timeout)
-            resp_data = resp.read()
+            resp_data = get_file_contents(
+                download_file(url, cache=cache, show_progress=False,
+                              timeout=data.conf.remote_timeout))
```
```suggestion
                download_file(url, cache=cache, show_progress=False))
```

`timeout` setting here is redundant, as it is already the default behavior:

https://github.com/astropy/astropy/blob/986d5b74145c4cf6ee47c303d709832d199df24f/astropy/utils/data.py#L1138-L1139

pllim commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -1913,7 +1918,14 @@ def from_name(cls, name, frame='icrs', parse=False):
+                             "you are using caching, i.e. if `cache=True`.")
+
+        if overwrite_cache and cache:
+            overwrite_cache = "update"
```
A bug?
```suggestion
            cache = "update"
```

pllim commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -1904,6 +1905,10 @@ def from_name(cls, name, frame='icrs', parse=False):
             sub-arcsecond accuracy for coordinates.
+        cache : bool, optional
+            Determines whether to cache the results or not.
+        overwrite_cache : bool, optional
```
I don't understand the need to have two different keywords to do this. Why not ask user to use `cache='update'` directly like in `get_icrs_coordinates` above?

adrn replied:
Sigh, fine. I just hate string arguments like this so much!!

mhvk replied:
Maybe we can slowly move to `Enum` - https://docs.python.org/3/library/enum.html

pllim commented on astropy/coordinates/tests/test_name_resolve.py:
```diff
@@ -136,6 +140,38 @@ def test_names():
+
+        icrs1 = get_icrs_coordinates("castor", cache=True)
+
+        # This is a weak test: we just check to see that a url is added to the
```
Why not have a stronger test with [astropy.utils.data.get_cached_urls](https://docs.astropy.org/en/latest/api/astropy.utils.data.get_cached_urls.html#astropy.utils.data.get_cached_urls)?

adrn replied:
Because I didn't know about that :) - I'll add a new check

pllim commented on astropy/coordinates/tests/test_name_resolve.py:
```diff
@@ -136,6 +140,38 @@ def test_names():
+        with shelve.open(urlmapfn) as url2hash:
+            assert len(url2hash) == 1
+
+        # Try reloading coordinates, now should just reload cached data:
```
I wonder if you can temporarily disable internet with [no_internet](https://github.com/astropy/pytest-remotedata/blob/371dd7d3f351fb0f672d1a7a9fb471ce63cbae07/pytest_remotedata/disable_internet.py#L172) to really test this. 🤔

adrn commented:
Question: Do we want the default for `from_name()` to be `cache=False` or `cache=True`?

adrn commented:
Thanks @pllim - I addressed your comments.

adrn commented:
Note to self: rebase when #10274 is merged (assuming that fixes the failing test)

mhvk commented:
@adrn - on whether to cache or not by default - my preference would be something that expires, but I don't think we have that yet! Failing that, I think `True` makes the most sense - the use case @eteq mentioned, of finding oneself without internet and wanting to continue to work, is I think one of the strongest ones. Plus, if one is using `from_name` anyway, clearly one is not too worried about small changes in position.

adrn commented:
Thanks @mhvk - I also think that `cache=True` makes sense as the default. I updated and pushed.

mhvk commented on CHANGES.rst:
```diff
@@ -40,6 +40,9 @@ astropy.coordinates
   change with future releases. [#10238]
 
+- The ``SkyCoord.from_name()`` and Sesame name resolving functionality now
+  optionally caches results locally. [#9162]
```
Darn - should remove "optionally" since it is the default. Maybe "now is able to cache results locally and will do so by default"

adrn replied:
Good point! Thanks for catching this

adrn commented:
@bsipocz This is not critical and could be moved to 4.2. On the other hand, I think it is ready to be merged now. I leave it up to you!

mhvk commented:
Well, my goal this morning was to empty the slate of 4.1-milestoned issues and PRs, so I'm going to continue to do so and merge... (hoping that I do more good than bad - though my inner dictator is having a blast, with the part of me that stifles it distracted by the beautiful weather - on a morning like this, with the sky pure blue, without brown haze on the horizon and unspoiled by con trails, the air pure and the ears filled with bird song and wind blowing throw new, light green leaves, and the noise smelling flowers rather than traffic - I am hoping some things of the lockdown will stay...).

adrn commented:
@mhvk OK, thanks. Yes, about to head outside myself - my eyes aren't focusing at infinity as much as they need to be! BTW: that message should be archived under #githubpoetry!

pllim commented:
@adrn -- I created a `#githubpoetry` channel on Slack...
